### ISSUE TITLE
Generic optional parameters are potentially dangerous

### ISSUE THREAD

**Issue by shicks** [google/closure-compiler#1176]
The following code should produce a warning, but it does not:

``` javascript
// externs
/**
 * @param {V} value
 * @param {R=} opt_return
 * @return {V|R|undefined}
 * @template V,R
 */
get = function() {};

// actual code
/** @type {number} */ var x = get('x');
```

Somehow, the unbound `R` in the return is convincing the compiler that it's okay to assign something that can clearly be `undefined` to a `number`.  There are other issues here, but ideally the `R` in the return should always be `undefined` unless the parameter is given, but there's no way to write that with the current type language.  This would allow tightening the return type of `goog.object.get`, for instance, to be simply `{V|R}` rather than `{V|R|undefined}`, and would also make it safer.

**Comment by dimvar** [comment]
That's a cool bug! Basically, we should initialize the type variable of a missing optional argument to undefined, not to unknown. Link to debugger repro:

http://closure-compiler-debugger.appspot.com/#input0%3D%250A%252F**%2520%2540type%2520%257Bnumber%257D%2520*%252F%2520var%2520x%2520%253D%2520f('asdf')%253B%26input1%26conformanceConfig%26externs%3D%252F**%250A%2520*%2520%2540param%2520%257BV%253D%257D%2520x%250A%2520*%2520%2540param%2520%257BR%253D%257D%2520y%250A%2520*%2520%2540return%2520%257BV%257CR%257Cundefined%257D%250A%2520*%2520%2540template%2520V%252CR%250A%2520*%252F%250Afunction%2520f(x%252C%2520y)%2520%257B%257D%253B%250A%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_TYPES_NEW_INFERENCE%3D1%26MISSING_PROPERTIES%3D1

**Comment by shicks** [comment]
I considered suggesting that, but wasn't sure if it would always work.  I can imagine cases like

    /**
     * @param {T=} opt_elem   // or even {...T} var_args
     * @return {Set<T>}
     * @template T
     */
    makeSingletonOrEmptySet = function(opt_elem) {
      const set = new Set();
      if (arguments.length) s.add(opt_elem);
      return set;
    };

In this case, we really can return any kind of set, since it's empty.

**Comment by dimvar** [comment]
Good point. I guess we can make the change, and if practice shows that it finds more actual bugs than false positives, we can keep it. Otherwise, we'll revert it and we'll miss a few true warnings.

**Comment by ChadKillingsworth** [comment]
We should fix this both directions.

```JavaScript
/** @param {number} foo */
function(foo, bar) {
  alert(bar); //bar is an unknown type
}
```
I've always thought a partially annotated function should be a lint warning at least.
